ECS Refactor.

Richardn2002 opened the pull request:
*Putting my plans and designs here later.*

Richardn2002 committed:
Foundational block of ECS: type-erased `Vec`.
See discussion at https://discord.com/channels/273534239310479360/1215712247654125588

Richardn2002 committed:
Finished readings on provenance.
https://www.ralfj.de/blog/2022/04/11/provenance-exposed.html

Richardn2002 committed:
`World` for hosting all nodes of different comps of a puppet.
Safe interfaces of `AnyVec` removed, as tested and usage constrained.

Richardn2002 committed:
Major ECS refactor.
Inochi2D node types into Inox2D components, see comments in `node/components.rs`.
`Puppet` as self-managed struct, reads INP file into a tree of nodes and a world of components to construct self.
No `Copy/Clone` on `InoxNode`, components and `JsonValue`, cutting Aka.inp loading time to 1/3.

Richardn2002 committed:
Absolute transforms of nodes as components.

Richardn2002 committed:
RenderCtx with components in mind. Init and reading methods.
Storing absolute transforms is no longer RenderCtx's job.
Put graphics buffer stuff to a seperate file.

Richardn2002 committed:
Componentize render contexts and store them in puppets.
Should be easy for incorporating future renderable node types.

Richardn2002 committed:
Complete render.rs rewrite.
Get required data from components instead.
Changes to the renderer traits:
- remove `render()`
- `on_begin_mask()` -> `on_begin_masks()`
- `set_mask_mode()` -> `on_begin_mask()`
- `draw_mesh_self()` -> `draw_textured_mesh_content()`
- `draw_part()` -> `draw_drawable()`

Richardn2002 committed:
Comments for `render.rs` and lighter interfaces.

Richardn2002 committed:
Pass `InoxNodeUuid` of the node in question to draw calls in case some backend gets smart.
For example reusing some results for a same node.

Richardn2002 committed:
Deformations by parameters.
Enum `Deform` and `DeformSrc` to get prepared for Meshgroup implementation.
One `DeformStack` per deform-able node for combining multiple deforms in a controlled way.
Drawable related stuff, especially the drawable kind determination logic, moved to its own places from `render.rs`, as apparently other systems (parameters for example) may want to know is a node a drawable.

Richardn2002 committed:
Components to store zsorts and transforms that are being modified across frames.
so that `InoxNode` stores constant initial values.

Richardn2002 committed:
zsort and deform update for `RenderCtx`.

Richardn2002 committed:
`Param` init and set.
The same reasoning behind `ctx` structs: Any data not in a model file is additional context data for certain functionalities, which should be only initialized upon demand.
End users only reset and set the values of params. When and how to actually apply params per frame is up to this library.
Removed `physics_ctx` placeholder that is present on local but should not be published.

Richardn2002 committed:
Implement reset methods for puppet to run across frames.
General pattern for a "system": Takes the form of an optional `-Ctx` struct attached to a puppet, which can be initialized if dependencies met.
Interactions with the puppet data are done in methods of the struct, so a `None` system can do nothing.
Current systems: Transform, Param, Render.

Richardn2002 committed:
`render-opengl` compiles.
With parameters and ECS system in place.
Changes to `trait InoxRenderer` during implementation:
- `new()` can have custom params.
- `on_begin_draw()` and `on_end_draw()`
Make `vao` creation, binding and unbinding, and vbo attaching, more explicit.
Move previous OpenGL `InoxRenderer` implementation into new framework.
Remove duplicate shader bind and uniform uploads during iteration over masks for a node (do once inside `on_begin_masks()`).

Richardn2002 committed:
Type annotations for `transmute()` by clippy.

Richardn2002 committed:
More comments on `upload_deforms_to_gl()`.

Richardn2002 committed:
Fix uninitialized zsort.

Richardn2002 committed:
Fix zsort, thus rendering is back for `Aka.inp`.
Properly implement zsort inheritance (+).
Correctly exclude composite children from `root_drawables_zsorted`.
Turn off param application temporarily, so this commit works for `Aka.inp`.

Richardn2002 committed:
Params (and deforms) back online: Dirty zero check patching for an old bug.
Currently, overflow will happen if value (0, 0) is applied on a param, which is clearly a bug.
The actual fix is not in the scope of ECS refactor.

Richardn2002 committed:
Physics cleanup.
The philosophy of the ECS refactor naturally leads to this cleanup:
In this refactor, spec-defined variables either goes into `InoxNode` or a component. If there are additional variables needed for the execution of a certain module, they are attached as additional components later.
It is revealed that bob position of a pendulum is such a variable.
After cleanup, all pendulum-like systems can provide methods for accessing and updating their bobs' position, implementing `trait Pendulum`.
And then `trait SimplePhysicsCtx` will be auto implemented for all `impl Pendulum`, handling the mapping from transforms to simulation input, from simulation input to parameter values.
More comments are added.

Richardn2002 committed:
Physics back online.
The implementation is inefficient (check changes in `puppet.rs`), yes, but clear, and the bottleneck is not here anyways.
Cloning of `SimplePhysics` is required for working around ownership problems, until dark magic implemented on `World` enabling multiple mutable handles to different components.

Speykious commented on inox2d/src/node/components.rs:
```diff
@@ -0,0 +1,31 @@
+/// Internal component storing zsort being determined in between frames.
+pub(crate) mod zsort;
+
+pub use composite::Composite;
+pub use drawable::Drawable;
+pub use simple_physics::SimplePhysics;
+pub use textured_mesh::TexturedMesh;
+
+pub(crate) use deform_stack::DeformStack;
+pub(crate) use transform_store::TransformStore;
+pub(crate) use zsort::ZSort;
```
Given the respective sizes of these modules, I think all of the structures defined in there should live in this single file instead. Right now it feels artificially sparse, the individual structures don't have any complex logic on their own that would be contained in separate files, making the modules shallow.

Richardn2002 replied:
This is a valid point, as component definition files indeed should only contain data and no logic.
And I should probably put `DeformStack` logic to somewhere else, likely under `render`, cause this is indeed *some* logic.

Speykious commented on inox2d/src/node/components/deform_stack.rs:
```diff
@@ -0,0 +1,81 @@
+
+/// Source of a deform.
+#[derive(Hash, PartialEq, Eq, Copy, Clone)]
+pub(crate) enum DeformSrc {
```
I think I'd prefer `DeformSource` as a name

Speykious commented on examples/render-opengl/src/main.rs:
```diff
@@ -41,12 +41,18 @@ fn main() -> Result<(), Box<dyn Error>> {
 		(model.puppet.meta.name.as_deref()).unwrap_or("<no puppet name specified in file>")
 	);
 
+	tracing::info!("Setting up puppet for transforms, params and rendering.");
+	model.puppet.init_transforms();
+	model.puppet.init_rendering();
+	model.puppet.init_params();
+	model.puppet.init_physics();
```
So if this is all called at the same time, why are they separate functions in the first place?

Richardn2002 replied:
Related to the comment below: It is actually possible to:
- Not call any of these init functions, only load whatever is in the inp/inx file into RAM. Potential applications include a simple parser.
- Only init transform + render, so displaying a static model is possible. Potential applications include a model previewer.
- Only init transform + render + param, to render a model without physics. Potential applications include a renderer with low perf settings.
- Init all and enjoy all.

Such modularized init is a nice byproduct of the refactor so I enable it.

Speykious commented on inox2d/src/node/drawables.rs:
```diff
@@ -0,0 +1,66 @@
+			Some(drawable) => drawable,
+			None => return None,
+		};
+		let transform = &comps
+			.get::<TransformStore>(id)
+			.expect("A drawble must have an associated transform.")
+			.absolute;
```
It's weird that `.get().expect()` is needed for `TransformStore`. Doesn't literally every node need to store transform data? Ideally this function should just get it from the `InoxNode` itself or something similar.

Richardn2002 replied:
See [above](https://github.com/Inochi2D/inox2d/pull/86#discussion_r1698746606).

Speykious commented on inox2d/src/node/drawables.rs:
```diff
@@ -0,0 +1,66 @@
+		let composite = comps.get::<Composite>(id);
+
+		match (textured_mesh.is_some(), composite.is_some()) {
+			(true, true) => panic!("The drawable has both TexturedMesh and Composite."),
+			(false, false) => panic!("The drawable has neither TexturedMesh nor Composite."),
```
A node having both a `TexturedMesh` and a `Composite` is something that is up to the model, so it's something that can actually happen. Crashing doesn't make sense here.
I propose that we return something that seems wrong instead of panicking. For example `(false, true) | (true, true) => Some(DrawableKind::Composite(...))` (which gives priority to the composite and ignores the textured mesh) and `(false, false) => None`.
Then somewhere else during initialization, we can log an error with tracing to warn the user of a malformed of corrupted model, and even go as far as to say that the textured mesh will be ignored.

Richardn2002 replied:
A node having both components is out of Inochi2D spec. It is us who choose to represent `class Part : Drawable` and `class Composite : Drawable` in the D ref impl with composition in Rust. If it is type wise not possible to describe a node with both (you can't override a same function twice!) in D, I guess it is not proper for Inox2D core to do so.

If you are persuaded that this is out of spec, and still wants to keep such behavior as Inox2D extension, then my opinion on such an extension is: Keep having both of these components a signal of a bug on our end or an improper usage from the end user. Use components of other types for representing such special nodes that act as both textured mesh and composite.

Speykious replied:
It being out of the spec got past me, but it is not really a good reason to add points of panic. inox2d is only responsible for moving puppets, it's weird and unexpected for such a library to crash the end user's program. In our case it's better to continue with a wrong-ish state because the end user can still run their application without it crashing, which would be more annoying. Think textures not loading being replaced by black-magenta default textures instead of blowing up and not running. So I maintain that if it has a problem it should ideally log it and move on as best it can.

Speykious commented on inox2d/src/node/drawables.rs:
```diff
@@ -0,0 +1,66 @@
+	/// Tries to construct a renderable node data pack from the World of components:
+	/// - `None` if node not renderable.
+	/// - Panicks if component combinations invalid.
+	pub(crate) fn new(id: InoxNodeUuid, comps: &'comps World) -> Option<Self> {
```
I'm not sure if this is the best place for this function or if it's even the right approach. It doesn't feel right to have to look at the components of a node only for it to return an enum that we have to then match against instead of doing it more directly. I don't have a nicer solution right now though but I'm probably gonna think about refactoring this part in the future.

Richardn2002 replied:
Related to [above](https://github.com/Inochi2D/inox2d/pull/86#discussion_r1698768218). This "component combination to enum variants" mapping is for following the Inochi2D spec. There should be a clear way of detecting "official" nodes, and then executing them in the "official" way.

Instead of removing this, what should be done further is a way to inject a end-user-provided routine to handle custom nodes, and before we hand over a node to a custom routine, we still need a way to extract the non-custom part and do what inox2d is supposed to do, which is what we have here, this `new()`.

Speykious commented on inox2d/src/node/components/drawable.rs:
```diff
@@ -0,0 +1,83 @@
+	SliceFromLower,
+}
+
+impl BlendMode {
+	pub const VALUES: [BlendMode; 7] = [
+		BlendMode::Normal,
+		BlendMode::Multiply,
+		BlendMode::ColorDodge,
+		BlendMode::LinearDodge,
+		BlendMode::Screen,
+		BlendMode::ClipToLower,
+		BlendMode::SliceFromLower,
+	];
+}
```
Why is this here? It's not even used :v
(my guess is you intended to use it for something but forgot to remove it)

Richardn2002 replied:
Fun fact: this is used by the WGPU renderer for some reason. Tho the WGPU renderer is seriously out of date now...

Speykious replied:
Let's keep it for now until we revive the WGPU renderer then. I'll probably move that to inox2d-wgpu eventually

Speykious commented on inox2d/src/render.rs:
```diff
@@ -1,378 +1,386 @@
+	pub(crate) fn reset(&mut self, nodes: &InoxNodeTree, comps: &mut World) {
+		for node in nodes.iter() {
+			if let Some(DrawableKind::TexturedMesh(..)) = DrawableKind::new(node.uuid, comps) {
+				let deform_stack = comps
+					.get_mut::<DeformStack>(node.uuid)
+					.expect("A TexturedMesh must have an associated DeformStack.");
```
I see a few of these `DeformStack` component fetches. The best thing to do here in case it doesn't exist is to do nothing. This is again a case where you'd log an error at initialization time if the puppet is wrong.

Richardn2002 replied:
My initial response would be: If rendering is initialized, then a `TexturedMesh` MUST have a `DeformStack`, because without a deform (outputted by the respective stack), the shader just don't work. This is not a sign of wrong puppet. This is a sign of wrong rendering initialization logic, which is our problem. So probably better to panic then to continue silently.

My second response: Oh, you give me the insight that we can NOT create a `DeformStack` for a `TexturedMesh` that is not deformed by any parameter (or, the parameter system is not initialized at all), and leave a default buffer region with zeroes for the GPU. In this case yeah we can just skip `None`.

Speykious commented on inox2d/src/render.rs:
```diff
@@ -1,378 +1,386 @@
-				node_render_ctx.trans = parent_trans * node_render_ctx.trans_offset.to_matrix();
-			}
-		}
+	/// Memory layout: `[[x, y], [x, y], ...]`
+	pub fn get_raw_verts(&self) -> &[f32] {
+		VertexBuffers::vec_vec2_as_vec_f32(&self.vertex_buffers.verts)
+	}
+	/// Memory layout: `[[x, y], [x, y], ...]`
+	pub fn get_raw_uvs(&self) -> &[f32] {
+		VertexBuffers::vec_vec2_as_vec_f32(&self.vertex_buffers.uvs)
+	}
+	/// Memory layout: `[[i0, i1, i2], [i0, i1, i2], ...]`
+	pub fn get_raw_indices(&self) -> &[u16] {
+		self.vertex_buffers.indices.as_slice()
+	}
+	/// Memory layout: `[[dx, dy], [dx, dy], ...]`
+	pub fn get_raw_deforms(&self) -> &[f32] {
+		VertexBuffers::vec_vec2_as_vec_f32(&self.vertex_buffers.deforms)
 	}
```
These getters aren't necessary. They add noise more than anything right now. Better just use the code that's inside them on the one call site where they're used and keep every field of the `RenderCtx` as public.

Richardn2002 replied:
Now I think it, such interfaces are loosely inspired by the "slab allocator" talk by Luna, where node data for GPU can be efficiently re-ordered for efficient batched rendering. In such case, actual data layout will be complex, but renderer implementers still want to see a contiguous memory region for upload. And this is where such "adapter" functions come in.

But true, we are so far away from actually having an idea of how to design such an allocator, so probably a good idea to remove these getters for now. We will have a better idea of what the interface should be when internal memory layouts really begin to change.

Speykious commented on inox2d/src/render.rs:
```diff
@@ -1,378 +1,386 @@
-	fn render(&self, puppet: &Puppet);
-	/// Finish one render pass.
-	fn on_end_scene(&self);
-	/// Actually make results visible, e.g. on a screen/texture.
-	fn draw_scene(&self);
+	/// Ref impl: Upload textures.
+	fn new(params: Self::NewParams, model: &Model) -> Result<Self, Self::Error>;
```
This function is not used polymorphically inside the library, so I think it's better to not specify any need for it. It'll remove the need for the `NewParams` associated type and make the `new` function of every renderer more flexible.

Speykious commented on inox2d/src/render.rs:
```diff
@@ -1,378 +1,386 @@
 	}
 }
+
+/// Dispatches draw calls for all nodes of `puppet`
+/// - with provided renderer implementation,
+/// - in Inochi2D standard defined order.
+///
+/// This does not guarantee the display of a puppet on screen due to these possible reasons:
+/// - Only provided `InoxRenderer` method implementations are called.
+/// For example, maybe the caller still need to transfer content from a texture buffer to the screen surface buffer.
+/// - The provided `InoxRender` implementation is wrong.
+/// - `puppet` here does not belong to the `model` this `renderer` is initialized with. This will likely result in panics for non-existent node uuids.
+pub fn draw<T: InoxRenderer>(renderer: &T, puppet: &Puppet) {
+	renderer.on_begin_draw(puppet);
+	renderer.draw(puppet);
+	renderer.on_end_draw(puppet);
+}
```
I don't think it's a good idea to define this function. Calling `on_begin_draw` then `draw` then `on_end_draw` is more easy to understand than calling a separate `inox2d::render::draw(...)` function which itself has to be inspected.

Richardn2002 replied:
This function is there for restricting the visibility of crate internals. I don't want the trait containing the auto implemented `.draw()` (currently under `InoxRendererCommon`) to be public for end users to `use` and get the `.draw()`. This gives a wrong idea of `InoxRendererCommon` can be implemented themselves.

As for the urge of inspecting everything, my response to suppressing that urge is to use intensive comments (something I had already done), so that users can start to believe as long as they implement what are prompted to implement, inox2d will do the rest, and do what they expect.

I will still admit that I am not an expert in all Rust features related to `trait`s, so feel free to offer alternatives controlling the visibility, or explain why visibility is not important here.

Speykious replied:
I think what we can do is to rename `InoxRendererCommon` to `InoxRendererExt` (`Ext` is a common suffix for extension crates), making it clear that it's an extension trait. That way, it being public will be no problem as it'll be clear that it's automatically added functionality for any renderer, and that you can't (and are not supposed to) implement any of that yourself.

Speykious commented on inox2d-opengl/src/gl_buffer.rs:
```diff
@@ -11,71 +9,50 @@ unsafe fn upload_array_to_gl<T>(gl: &glow::Context, array: &[T], target: u32, us
+///
+/// This function will return an error if it couldn't create a vertex array.
+///
+/// # Safety
+///
+/// No prerequisites.
+pub unsafe fn setup_gl_buffers(
```
If there are no prerequisites, I guess the function shouldn't be `unsafe` and it should just have an `unsafe` block inside.

Speykious reviewed (changes requested):
I think overall I like this change. There's a few things here and there that I commented on but otherwise it's nice. I'll want to do another refactor pass on this when it's finalized and merged to improve on the doc comments you provided, and then later see what I can do about this very hacky `AnyVec` implementation. I'm probably going to replace it with an actual page-allocated arena, if all our supported targets allow for it.

Richardn2002 commented:
What does "another refactor pass" mean :eyes:? IMO we really should be doing MeshGroups and other rendering backends and even game engine integrations afterwards.
Sure the `World` implementation can be improved. I would like more efficient iteration, and, if possible, dark magic allowing obtaining two mutable handles to two components of different types, or even just two different components. But anyways I don't think reimplementing `World` would break too much stuff and cause a refactor.
The idea of "systems" and the process of systems applying to node data in order can be formalized, yes. If you want to call this a "refactor" I am also fine with it.

Btw, my `AnyVec` is hacky, intentionally, because I want to steal all the work done by Rust `std` team for `Vec`. We all have seen some perf results and we all know matching `std` performance is not easy :eyes:. And it is not that `Vec` can't use a custom allocator. I do suggest giving more thoughts to the idea of reusing `Vec` as I am doing now.

Richardn2002 committed:
`DeformSrc` -> `DeformSource`
And suppress variant unused warning... until Meshgroups.

Richardn2002 committed:
Put all components into `compoents.rs`.
Responding to https://github.com/Inochi2D/inox2d/pull/86#discussion_r1698615893
Rename `ParamMapMode` to `PhysicsParamMapMode` for this change for clarity.

Richardn2002 committed:
More direct `VertexBuffers` access.
In response to https://github.com/Inochi2D/inox2d/pull/86#discussion_r1698693637

Richardn2002 committed:
Better safety specifications related to GL buffer operations.
In response to https://github.com/Inochi2D/inox2d/pull/86#discussion_r1698729703
Plus more robust `upload_array_to_gl()`, with more annotations inside.

Richardn2002 committed:
Remove `new()` from `trait InoxRenderer`.
In response to https://github.com/Inochi2D/inox2d/pull/86#discussion_r1698697131
...and why `Sized` is required in the first place?

Richardn2002 committed:
Better `on_begin_draw()` `on_end_draw()` `draw()` interfaces.
In response to https://github.com/Inochi2D/inox2d/pull/86#discussion_r1698705645
The core crate does not need to know about the presence of stuff to do before and after `draw()`. This is up to renderer implementations.

Richardn2002 committed:
`tracing::warn!()` instead of `panic!()` on non-standard component combinations for a Drawable.
In response to https://github.com/Inochi2D/inox2d/pull/86#discussion_r1700035650

Richardn2002 committed:
Do not give `TexturedMesh` that is not going to be deformed by any source a `DeformStack`.
Thus accordingly, fewer `DeformStack` resetting and reading will happen.
Optimization inspired by https://github.com/Inochi2D/inox2d/pull/86#discussion_r1698688185

Richardn2002 committed:
Generalize `Mesh` into a standalone component.
For future meshgroup implementation.

Speykious committed:
Move `OpenglRenderer::new` up in impl block

Speykious commented:
Looks good to me. Now we just need to revive inox-wgpu

Speykious commented:
After discussing it on Discord, we decided it was best to get rid of the WGPU renderer for now. Neither of us can maintain it because we don't know WGPU at all, and once either of us learns it, we're gonna need to rewrite it from scratch anyway because of this refactor and possble future ones to come.

Speykious committed:
Merge branch 'main' into pr/Richardn2002/86

Speykious committed:
Fix WebGL example

Speykious reviewed (approved):
We're good now
